trace/config.go

```diff
@@ -16,25 +16,72 @@ package trace
 
 import "go.opencensus.io/trace/internal"
 
+// Default limits for the number of attributes, message events and links on each span
+// in order to prevent unbounded memory increase for long-running spans.
+// These defaults can be overriden with trace.ApplyConfig.
+// These defaults can also be overriden per-span by using trace.StartOptions
+// when creating a new span.
+// TODO: Add an annnoation limit when priorities are implemented.
+const (
+	DefaultMaxAttributes    = 32
+	DefaultMaxMessageEvents = 128
+	DefaultMaxLinks         = 32
+)
+
 // Config represents the global tracing configuration.
 type Config struct {
 	// DefaultSampler is the default sampler used when creating new spans.
 	DefaultSampler Sampler
 
 	// IDGenerator is for internal use only.
 	IDGenerator internal.IDGenerator
+
+	// The below config options must be set with a GlobalOption.
+	// maxAttributes sets a global limit on the number of attributes.
+	maxAttributes int
+	// maxMessageEvents sets a global limit on the number of message events.
+	maxMessageEvents int
+	// maxLinks sets a global limit on the number of links.
+	maxLinks int
+}
+
+// GlobalOption apply changes to global tracing configuration.
+type GlobalOption func(*Config)
+
+// WithDefaultMaxAttributes sets the default limit on the number of attributes.
+func WithDefaultMaxAttributes(limit int) GlobalOption {
+	return func(c *Config) {
+		c.maxAttributes = limit
+	}
+}
+
+// WithDefaultMaxMessageEvents sets the default limit on the number of message events.
+func WithDefaultMaxMessageEvents(limit int) GlobalOption {
+	return func(c *Config) {
+		c.maxMessageEvents = limit
+	}
+}
+
+// WithDefaultMaxLinks sets the default limit on the number of links.
+func WithDefaultMaxLinks(limit int) GlobalOption {
+	return func(c *Config) {
+		c.maxLinks = limit
+	}
 }
 
 // ApplyConfig applies changes to the global tracing configuration.
 //
 // Fields not provided in the given config are going to be preserved.
-func ApplyConfig(cfg Config) {
+func ApplyConfig(cfg Config, o ...GlobalOption) {
```

Comment:
This is where it gets tricky:

With the current trace.ApplyConfig implementation:
1) cfg is a required parameter and is expected to be a partial update (preserving what was already set).
2) We want limits set to <= 0 to mean "disabled".
3) Go's zero value for an int is 0, so if someone changes one limit (or even just the DefaultSampler), the others would incorrectly become disabled.

To handle this, I added in a new variadic GlobalOption (which doesn't break the API), but I think the cfg arg should be deprecated in favor of all variadic options. I'm sure a lot of clients currently use this, though.

Another option: we could keep the cfg struct arg, add *int fields in for the limits, and add a trace.Limit function that takes an int and returns an *int, but I don't like that API as much.